src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/NetCorePackageReferenceProject.cs

```diff
@@ -214,40 +217,111 @@ private void IgnoreUnsupportProjectReference(PackageSpec project)
 
-        public override Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(CancellationToken token)
+        public async override Task<IEnumerable<PackageReference>> GetInstalledPackagesAsync(CancellationToken token)
         {
-            PackageReference[] installedPackages;
+            var lockFilePath = await GetAssetsFilePathAsync();
+            var lockFileFormat = new LockFileFormat();
 
             var packageSpec = GetPackageSpec();
-            if (packageSpec != null)
-            {
-                installedPackages = GetPackageReferences(packageSpec);
-            }
-            else
+
+            var fileInfo = new FileInfo(lockFilePath);
+            if(fileInfo.LastWriteTimeUtc > _lasTimeAssetsModified)
             {
-                installedPackages = new PackageReference[0];
-            }
+                var lockFile = lockFileFormat.Read(lockFilePath);
+                var packageSpecAsset = lockFile.PackageSpec;
+                if (packageSpec != null)
+                {
+                    _installedPackages = GetPackageReferences(packageSpec, packageSpecAsset, lockFile.Targets);
+                }
+                else
+                {
+                    _installedPackages = new PackageReference[0];
+                }
+
+                _lasTimeAssetsModified = fileInfo.LastWriteTimeUtc;
+            } 
 
-            return Task.FromResult<IEnumerable<PackageReference>>(installedPackages);
+            return _installedPackages;
         }
 
-        private static PackageReference[] GetPackageReferences(PackageSpec packageSpec)
+        private static PackageReference[] GetPackageReferences(PackageSpec packageSpec, PackageSpec assetsPackageSpec, IList<LockFileTarget> targets = null)
         {
             var frameworkSorter = new NuGetFrameworkSorter();
 
+            PackageSpec combinedPackageSpec = new PackageSpec();
+
             return packageSpec
-                .TargetFrameworks
-                .SelectMany(f => GetPackageReferences(f.Dependencies, f.FrameworkName))
-                .GroupBy(p => p.PackageIdentity)
-                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
-                .ToArray();
+               .TargetFrameworks
+               .SelectMany(f => GetPackageReferences(f.Dependencies, f.FrameworkName, assetsPackageSpec, targets))
+               .GroupBy(p => p.PackageIdentity)
+               .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+               .ToArray();
         }
 
-        private static IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework)
+        private static IEnumerable<PackageReference> GetPackageReferences(IEnumerable<LibraryDependency> libraries, NuGetFramework targetFramework, PackageSpec assetsPackageSpec, IList<LockFileTarget> targets)
         {
+            TargetFrameworkInformation packageSpec = default;
```

Comment:
The naming is confusing here. 
TargetFrameworkInformation is not a package spec.